ACPs/77-reinventing-subnets/README.md

```diff
@@ -416,6 +414,90 @@ class ValidatorQueue:
+
+#### Block Timestamp Validity Change
+
+To easily track $x$ between blocks, the difference between $V$ and $T$ must be constant in the $\Delta t$ between blocks. $x$ can then simply be updated to $\max(x + \Delta t \cdot (V - T), 0)$ after each valid block. If $V-T$ is not constant between blocks, this optimization cannot be made.
+
+Suppose the network currently has `11000` Subnet Validators. With $T$ being `10000`, $x$ will increase by $V-T$ or `1000` every second. If the next block's timestamp is `5` seconds after the current block's timestamp. $x$ will increase by `5000`. This is expected behavior _unless_ Subnet Validators should have been removed within those `5` seconds. If `100` Subnet Validators would exhaust their balance `1` second after the current block timestamp, $x$ should only increase by `900` every second after the first second elapses. $x$ should have increased by `1000` + `4 * 900` = `4500`, not `5000`.
+
+To ensure that the correct fee is always charged per unit of time, blocks are considered valid if their timestamps are no greater than the time at which the first Subnet Validator gets removed from a lack of funds. This upholds the invariant that the number of Subnet Validators remains constant between blocks.
+
+The block building protocol is modified to account for this change by first checking if the wall clock time removes any Subnet Validator due to a lack of funds. If the wall clock time does not remove any Subnet Validators, the wall clock time is used to build the block. If it does, a binary search may be performed between the parent block's timestamp and the wall clock time to find the time at which the first Subnet Validator gets removed.
+
+#### Fee Calculation
+
+Since $V-T$ is guaranteed to be constant between blocks, a simple formula can be derived. Prior to processing the next block, the total Subnet Validator fee assessed in the $\Delta t$ between the current block and the next block is::
+
+$$\sum_{n=1}^{\Delta t} M \cdot \exp \left(\frac{x+n\cdot(V-T)}{K}\right)$$
+
+However, this formula does not hold when $V-T < 0$ and $x < \Delta t \cdot (V-T)$. This is because the $x$ is nonnegative according to its update function $x = \max(x + \Delta t \cdot (V - T), 0)$. The above formula can be used during the time period where $x$ is decreasing by $T-V$ and is not floored to $0$:
+
+$$\min\left(\Delta t, \left\lfloor{\frac{x}{T-V}}\right\rfloor\right)$$
+
+For the remaining time period where $x$ would be floored to $0$, the fee is:
+
+$$M \cdot \max\left(\Delta t - \left\lfloor{\frac{x}{T-V}}\right\rfloor, 0\right)$$
+
+After the Subnet Validator fee is calculated, $x$ can be updated:
+
+$$x = \max(x + \Delta t \cdot (V - T), 0)$$
```

Comment:
```suggestion
Prior to processing the next block, the total Subnet Validator fee assessed in the $\Delta t$ between the current block and the next block is:

```python
# Calculate the fee to charge over Δt
def cost_over_time(V:int, T:int, x:int, Δt: int) -> int:
    cost = 0
    for _ in range(Δt):
        x = max(x + V - T, 0)
        cost += fake_exponential(M, x, K)
    return cost
```

Reply:
I couldn't figure out how to have the suggested change work correctly... But there needs to be a trailing "```" in the above suggested diff